Snaps documentation (#331)
MetaMask/metamask-docs: docs/guide/snaps-introduction.md

> # Snaps Tutorial

Reviewer: This does not seem like a tutorial. Is this a mistake?

Reviewer: Since this page is named "Snaps Introduction" and there is a subheader below called "Snaps Tutorial", this should probably be:

```suggestion
# Snaps Introduction
```

## What is Snaps?

> Snaps are a new evolution of MetaMask's functionality. By using MetaMask Flask in conjunction with the Snaps API, developers are able to bring their own code to the wallet in a custom-made code laboratory. MetaMask is the first wallet to provide this openness and flexibility to developers, paving the way to scale the number of features and personalization available to the wallet and your Dapp.

Reviewer: In the first three sentences, we've described snaps as "a new evolution", and as a "custom-made code laboratory", and congratulated ourselves on being the first to create such a thing, all without saying what it is. This does not seem helpful. We should be explaining what snaps is, not trying to sell it and the wallet itself.

Reviewer 2: So, what do you propose? Do we cut this copy entirely?

Reviewer: This is what I have so far for this section:

>Snaps is a system that allows safely expanding the capabilities MetaMask. A "snap" is a program that we run in an isolated environment that can customize the wallet experience in some way. For example, a snap can add new APIs to MetaMask, add new features, or modify existing functionality using internal APIs. Snaps makes extending MetaMask easier and safer, for both users and developers.

Reviewer 2: I'm OK with this if you guys are.

> Snaps allows developers to build and plug in their own APIs to MetaMask. No longer do you have to rely on MetaMask to roll out features and updates. The rate of innovation, and MetaMask's flexibility, are limited only by your imagination and your capacity to build. 

Reviewer: The first sentence is good except it kinda sells snaps short a bit. Building APIs doesn't quite encompass things like enhancements to the wallet itself or integrating with internal MetaMask APIs.

The part about not having to rely on MetaMask to roll out new features and updates doesn't ring true to me. I get what you were going for, that one of the goals of snaps is to allow developers to innovate faster than our team alone could hope to. But maybe we can phrase this more carefully. Snap developers will still very much be dependent upon us and our release schedule.

The last sentence doesn't ring true either :sweat_smile: There remain MANY limitations to what developers can do with snaps. They run in a very restricted environment and the API they are given is quite small right now. Frankly it's probably the most restrictive environment most snap devs will have ever encountered.

Reviewer 2: -I wasn't sure how much of that we were going to push, and how soon; hence I left it deliberately vague
-There was copy you found elsewhere--about how we'll use Flask to get a better idea of what devs want, and incorporate those things faster--that might resolve your doubts here.
-How about we nix that last sentence, then?

Reviewer: Sorry for the fractured conversation - was thinking that [the draft here](https://github.com/MetaMask/metamask-docs/pull/331#discussion_r786326845) could replace both of these paragraphs. Do you think we need anything from this one?

I feel like there is still _something_ to say about the rate of innovation but I couldn't think of how to say it. Maybe the Flask description gets close to it. It's not an exact parallel because Flask is basically just for allowing experimentation and gaining feedback to improve the snaps system, whereas with snaps the ultimate goal isn't for all of the good ideas to get absorbed into MetaMask itself. The good ideas can stay as snaps, that's fine. But there is a parallel in that the systems are both meant to drive innovation, and they work together toward that goal.

But I don't know how we should say any of that here, and it seems fine if we omit it for now.

## MetaMask Snaps: Overview

> Broadly speaking, MetaMask Snaps consists of two things: (1) a way to run untrusted JavaScript inside the MetaMask application, and (2) APIs for websites and MetaMask to communicate with individual snaps. As with MetaMask’s [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html), communication occurs via JSON-RPC requests and responses.

Reviewer: I find the division of snaps into "two things" here as a very odd way to frame things.

I would describe snaps as "a way to run untrusted JavaScript inside the MetaMask application", that part is good. Then maybe we could go on to explain the API that we offer to snaps, and that we offer APis that allow websites to communicate with snaps. I can give this a shot later.

Reviewer: On further consideration I think it was just the specific wording here that was throwing me off. Generally this isn't bad.

Here's what I came up with trying to re-write it. Not sure that it's better though.

>Broadly speaking, MetaMask Snaps is a system that allows running untrusted programs inside MetaMask and granting them specific capabilities. Snaps are run in an isolated environment where they have access to a limited set of capabilities, determined by the permissions they were granted by the user during installation. As with MetaMask’s [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html), snaps communicate with MetaMask using a JSON-RPC API.
>
>New RPC methods have been added to our [Ethereum JavaScript provider API](https://docs.metamask.io/guide/ethereum-provider.html) as well. These methods allow websites to request that snaps be installed, and to communicate with individual snaps.

Reviewer 2: > On further consideration I think it was just the specific wording here that was throwing me off. 

I have this experience a lot when collaborating on a text, so I feel you

Your version gives a lot more detail, and given that we're writing for devs here to explain how to do something complex, that seems appropriate to me.

### Current functionality

> At present, snaps can make JSON-RPC requests of MetaMask, and websites can make JSON-RPC requests of MetaMask and snaps running inside MetaMask. In this way, snaps can expand MetaMask’s RPC API and change the behavior of the MetaMask application at runtime. Over time, MetaMask will use this to support different blockchains and other decentralized protocols, novel kinds of cryptography, novel kinds of crypto assets, and a countless number of other features.

Reviewer: This makes it seem like the ability for snaps to extend the RPC API is what will enable decentralized protocols, novel cryptography, etc. That's not really true. Primarily it will be us building snap APIs that allow customizing the wallet experience that will enable that stuff. Extending the RPC API is just one thing a snap can do among many things.

Also we should probably not be framing "novel kinds of cryptography" as a future feature. We will enable that at launch.

Reviewer: Rather than updating this to be consistent with the reference docs, maybe we can delete this entire section and refer there instead?

Reviewer: Or if we want to keep it, here's my current draft:

>At present, snaps can create new RPC methods for websites to call, they can call many of the same RPC methods that websites can call, and they have access to a limited set of snap-exclusive RPC methods (see the Snaps RPC API documentation for more information). Over time, MetaMask will expose more internal functionality as RPC methods, granting more capabililities to snaps.

Reviewer 2: I think deleting and referring the users directly to the Snaps API section of the MM RPC API docs sounds sensible. No need to frame things differently in different places.

That is what you mean by 'refer there instead', right?

Reviewer: Yep! OK.

> For the prototype Snaps system, snaps cannot modify the MetaMask UI, but can extend the MetaMask RPC API, and exchange arbitrary messages with websites visited by the user. Therefore, the user interface for any snap must exist entirely in the website during the prototype stage.

Reviewer: We might want to rephrase this as "Currently snaps cannot modify the MetaMask UI" rather than talking about a "prototype snaps system". I don't know what that could refer to other than Flask, and Flask will definitely support custom UI.

Reviewer: Here's my current draft:

>At the moment, snaps cannot modify the MetaMask UI directly. If a snap needs a UI, that UI must exist entirely on a website that can communicate with the snap. We will be allowing UI customization in future updates.

Reviewer 2: That's fine--again, I was downplaying the UI modification capabilities intentionally.

### Execution Environment

> As stated, snaps are untrusted JavaScript programs, but execute safely inside the MetaMask application. To isolate snaps from the rest of the application, and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance, and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects, e.g. the Promise constructor.

Reviewer: Nit: Reads better with the "As stated" omitted. Readers might be coming directly to this section rather than reading chronologically anyway.

```suggestion
Snaps are untrusted JavaScript programs, but execute safely inside the MetaMask application. To isolate snaps from the rest of the application, and to provide a “fully virtualizable” execution environment, MetaMask uses [Secure EcmaScript (SES)](https://github.com/endojs/endo/tree/master/packages/ses), a subset of JavaScript developed by [Agoric](https://agoric.com/). Among other things, SES allows us to restrict access to global JavaScript APIs and to isolate untrusted code from other parts of the application. SES does this at the cost of some performance, and incompatibility with some JavaScript practices, such as modifying prototypes of intrinsic objects, e.g. the Promise constructor.
```

## Sample snap tutorial

There are a number of simple snaps available for you to learn from and begin to get an idea of what your own snap needs. The following tutorial will walk you through a Hello World snap.

### Prerequisites

* nodejs `^14.15`
* nvm (recommended)
* git
* yarn v1 
* macOS, Linux, or Windows Subsystem Linux
  
### Getting Flask running in your local environment

> Flask, as described [here](LINK TO FLASK PAGE REQUIRED HERE), is a distribution channel for developmental, experimental, kooky stuff for MetaMask, and Snaps is the first feature to be released through it. 

Reviewer: We should update this description to match the latest draft, this "kooky stuff" one was replaced.

Reviewer: This is the blurb I was thinking of:

>Flask is a distribution channel, or playground, for developers that gives them access to additional unstable APIs. The goal of Flask is to maximize developer control, so that we can learn the full extent of what developers want to do with MetaMask, and may consider incorporating those lessons into the main MetaMask distribution.

Reviewer 2: Yeah, that's fine--I had thought we had gotten rid of the 'kooky stuff' copy throughout the resources, despite, you know, loving it :-P

> * Start by downloading [Flask](LINK HERE) and adding it to Chrome

Reviewer: We have the link now!
```suggestion
* Start by downloading [Flask](https://chrome.google.com/webstore/detail/metamask-flask-developmen/ljfoeinjpaedjfecbmggjgodbgkmjkjk) and adding it to Chrome
```
